### PR TITLE
feat: add `Remove-vRAGroup` function

### PR DESCRIPTION
**Pull request by rnelson0**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/blob/main/CONTRIBUTING.md) for making a pull request.

**Summary of Pull Request**

Add `Remove-vRAGroup`, similar to `Undo-vRAGroup`, but for non-VCF instances

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.
      Please describe:

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### CODE REVIEW

**Comment by rnelson0** [comment]
Bump, any updates on this?

**Comment by jshiplett** [comment]
Don't take this as a final decision as we haven't talked about it (as far as I know) yet, but here's my take:

The primary driver of this project is enabling these solutions on top of VCF. All of the pre-checks, etc., rely on an SDDC Manager appliance being present.

Given that context, adding a non-VCF function by duplicating an existing function and removing all of the SDDC Manager bits doesn't make a whole lot of sense to me. I'm willing to be convinced otherwise, but this seems like something that would be more suitable to a forked project.

Don't get me wrong - I think there is real value in this work. I just don't think it's right for this project.

Thoughts @tenthirtyam @GaryJBlake @rnelson0?

**Comment by rnelson0** [comment]
I can't speak for the ultimate decision here, but there is already a mix of VCF and non-VCF functions present. As some of the SDDC functions sit on top of the non-SDDC portions of vRA, it makes sense. This could be split into two portions, some VCF functions that depend on non-VCF functions and the non-VCF functions, but that would be difficult to split and difficult/tedious to maintain. That said, it's not like that's out of the norm, so I'll support whatever works - in fact, I don't need the VCF bits so it would be simpler for me, even if it's not for others!

**Comment by jshiplett** [comment]
Those are helper functions, written to do a specific product-centric thing, that are called in the higher-level functions like Get-vRAGroup, Remove-vRAGroupRoles, etc., are to Undo-vRAGroup. They are typically very compact and have little or no error checking built in. 

Upon closer inspection, it looks like it's going to fail out of the gate since you won't normally have a vRA auth token (inside $vraHeaders) active in your PowerShell session. Each of the helper functions here make direct REST API calls to a vRA appliance and assume that $vraHeaders is there and contains a valid bearer token. The Test-vRAAuthentication function in Undo-vRAGroup calls Request-vRAToken to instantiate the token and header variable.

**Comment by rnelson0** [comment]
I'm not sure I understand why `$vraHeaders` is an issue, `New-vRAGroup` expects that to be present as well. Would you like the functional aspect behind an `if (Test-vRAAuthentication -server $vcfVraDetails.loadBalancerFqdn -user $vraUser -pass $vraPass)` block instead?

**Comment by tenthirtyam** [comment]
I'll take a gander at this one next week: I think there's some opportunities to have methods of VCF-mode vs non.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
I've been noodling this one (_yes, finally_) and I don't have any severe concerns about merging this in since Rob's taken the time to propose an option for a use-case that we may actually see as vRealize Suite becomes more downstream vs upstream connected.

**_However_**, one option that we should experiment with here is by instead creating a new cmdlet. modify the existing `Undo-vRAGroup` to take an optional override.

 _Whoa, wut?_

For example, what if we instead update`Undo-vRAGroup` to the following high-level concept:

- By default, assume vRA is integrated in SDDC Manager (so we don't have breaking changes)
- Optionally, provide an override switch (_e.g._, `-direct`, _or something more accurate_) to use the direct path.
- If direct path, then the `-server`, `-user`, and `-password` parameters are not required to return data (user/pass/fqdn) from SDDC Manager, but the addition of `-vraFqdn` would be required.

Default:

```powershell
.EXAMPLE
Undo-vRAGroup -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -vraUser configadmin -vraPass VMw@re1! -displayName gg-vra-cloud-assembly-admins@rainpole.io
This example removes a group from vRealize Automation by displayName If Realize Automation is registered with SDDC Manager.
```

Direct:

```powershell
 .EXAMPLE
Undo-vRAGroup -direct -vraFqdn xint-vra.rainpole.io -vraUser configadmin -vraPass VMw@re1! -displayName gg-vra-cloud-assembly-admins@rainpole.io
This example removes a group from vRealize Automation by displayName if vRealize Automation is not registered with SDDC Manager.
```

This would likely require the use of parameter sets.

If this pattern could be established, we should then be able to apply the same of other vRealize-focued cmdlets and provide some much needed PowerShell-based automation to an even wider audience.

Thoughts?

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Review by tenthirtyam — COMMENTED** [review]
Refer to https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/pull/15#issuecomment-1235762133.

**Comment by rnelson0** [comment]
-Direct is good. I can look into the other changes as well, but I’m in vacation next week, so it will be at least 10 days.

**Comment by tenthirtyam** [comment]
Enjoy your time off! DM me if you have any questions when you're back.

🌴 

Ryan

**Comment by rnelson0** [comment]
I haven't forgotten about this, as with most of us there's a rush to get things in before the holiday season begins, but I should have some time in the next week or two to work on this.

**Comment by rnelson0** [comment]
@tenthirtyam I think this is finally ready for review. Apologies for the significant delays!

I was able to test this on my vRA instance. I do not have access to vCF to validate there. Though I did not modify that code path aside from indentation, I did add the parameter set and a branch around it, so it wouldn't hurt to have someone with vCF access validate it just in case.

Thanks!

File: PowerValidatedSolutions.psm1
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:1]: I think it may be better to assume the vcf-mode by default and only go direct with switch.

**Comment by rnelson0** [comment]
Poke to keep it alive.

**Comment by rnelson0** [comment]
Checking for status on this. It looks like I'm going to have to rebase, but before I go through that I wanted to see if there were any other changes required. Thanks @tenthirtyam!

**Comment by tenthirtyam** [comment]
Rebased for review next week. :)

**Comment by tenthirtyam** [comment]
Rob (@rnelson0 ) - I'm going to close this one for now but we may be revisitng this to ensure that the cmdlets will run for environments that have both deployment scenarios.

**Pull request closed without merging**